### DISCUSSION TITLE
Handling Asynchronous Operations in a For Loop with Promises in JavaScript

### DISCUSSION THREAD

**Discussion by ajmiraajmi** [community/community#139337] [Programming Help] [answered]
### Body

I'm facing an issue while working with asynchronous operations inside a for loop in JavaScript. I have a list of items, and for each item, I need to perform an asynchronous task (e.g., an API call). However, the loop completes execution before the asynchronous operations are finished, causing unexpected results.

### Code:

```javascript
const items = [1, 2, 3, 4, 5];

function asyncTask(item) {
  return new Promise((resolve) => {
    setTimeout(() => {
      console.log(Processed item: ${item});
      resolve(item);
    }, 1000);
  });
}

function processItems() {
  for (let i = 0; i < items.length; i++) {
    asyncTask(items[i]);
  }
}

processItems();

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by sabbirsami** [comment, marked as answer]
# Solution: Handling Asynchronous Operations in a For Loop with Promises

## Problem Overview

When working with asynchronous operations in JavaScript, running them inside a `for` loop can cause all asynchronous tasks to execute simultaneously. This leads to unexpected results, as the loop doesn't wait for the tasks to complete before moving to the next iteration.

In this case, you may want the tasks to execute one after the other, ensuring that each asynchronous operation finishes before the next one starts.

## Solution

To handle asynchronous tasks sequentially inside a loop, we can use `async`/`await` along with a `for...of` loop. This method ensures that the loop waits for each asynchronous task to complete before proceeding to the next iteration.

### Example Code (Updated Solution):

```javascript
const items = [1, 2, 3, 4, 5];

function asyncTask(item) {
  return new Promise((resolve) => {
    setTimeout(() => {
      console.log(`Processed item: ${item}`);
      resolve(item);
    }, 1000);
  });
}

async function processItems() {
  for (const item of items) {
    await asyncTask(item);
  }
}

processItems();
